contrib/node/src/python/pants/contrib/node/subsystems/resolvers/node_resolver_base.py

```diff
@@ -35,3 +36,25 @@ def _copy_sources(self, target, results_dir):
       dest = os.path.join(results_dir, os.path.relpath(source, source_relative_to))
       safe_mkdir(os.path.dirname(dest))
       shutil.copyfile(os.path.join(buildroot, source), dest)
+
+  def _get_target_from_package_name(self, target, package_name, file_path):
+    """Get a target from its dependency tree given a package name and relative file path.
+
+    This will only traverse direct dependencies of the passed target. It is not necessary
+    to traverse further than that because transitive dependencies will be resolved under the
+    direct dependencies and every direct dependencies is symlinked to the target.
+
+    Returns `None` if the target does not exist.
+
+    :param NodePackage target: A subclass of NodePackage
+    :param string package_name: A package.json name that is required to be the same as the target name
+    :param string file_path: Relative filepath from target to the package in the format 'file:<address_path>'
+    """
+    pattern = re.compile('^file:(.*)$')
+    address_path = pattern.match(file_path).group(1)
```

Comment:
Is it always necessary for these to be relative? Does `package.json` support having file urls that are relative to the repository root?

Pants somewhat-intentionally does not support relative addresses in BUILD files, because it makes it much, much harder to refactor them.

Reply:
Unfortunately no, the `file:` specifier is relative to the root of the `package.json` to be backwards compatible with "yarn" and "npm". Alternatively we could in the future add a new dependency type `pantsDependencies` as a potential follow up to this implementation which would take a pants target address which would be relative to the build root.